File: src/main/java/client/player/quest/requirement/Requirement.java
package client.player.quest.requirement;

import client.player.quest.QuestRequirementType;
import util.packet.PacketReader;

import java.util.HashMap;
import java.util.Map;

public abstract class Requirement {

    private int minLevel, npc;
    private final Map<Integer, Short> items = new HashMap<>();
    private final Map<Integer, Byte> quests = new HashMap<>();

    public int getMinLevel() {
        return minLevel;
    }

    public int getNpc() {
        return npc;
    }

    public Map<Integer, Short> getItems() {
        return items;
    }

    public Map<Integer, Byte> getQuests() {
        return quests;
    }

    public void setMinLevel(int minLevel) {
        this.minLevel = minLevel;
    }

    public void setNpc(int npc) {
        this.npc = npc;
    }

    public void decode(int flags, PacketReader reader) {
        if (containsFlag(flags, QuestRequirementType.MIN_LEVEL)) minLevel = reader.readShort();
        if (containsFlag(flags, QuestRequirementType.NPC)) npc = reader.readInteger();
        if (containsFlag(flags, QuestRequirementType.ITEM)) {
            int size = reader.readShort();
            for (int i = 0; i < size; i++) {
                items.put(reader.readInteger(), reader.readShort());
            }
        }
        if (containsFlag(flags, QuestRequirementType.QUEST)) {
            int size = reader.readShort();
            for (int i = 0; i < size; i++) {
                quests.put(reader.readInteger(), reader.readByte());
            }
        }
    }

    public boolean containsFlag(int flags, QuestRequirementType flag) {
        return (flags & flag.getValue()) == flag.getValue();
    }
}

File: src/main/java/client/player/quest/requirement/Requirement.kt
package client.player.quest.requirement

import client.player.quest.QuestRequirementType
import util.packet.PacketReader

abstract class Requirement {

    var minLevel = 0
    var npc = 0
    val items: MutableMap<Int, Short> = HashMap()
    val quests: MutableMap<Int, Byte> = HashMap()

    open fun decode(flags: Int, reader: PacketReader) {
        if (containsFlag(flags, QuestRequirementType.MIN_LEVEL)) minLevel = reader.readShort().toInt()
        if (containsFlag(flags, QuestRequirementType.NPC)) npc = reader.readInteger()
        if (containsFlag(flags, QuestRequirementType.ITEM)) {
            val size = reader.readShort().toInt()
            for (i in 0 until size) {
                items[reader.readInteger()] = reader.readShort()
            }
        }
        if (containsFlag(flags, QuestRequirementType.QUEST)) {
            val size = reader.readShort().toInt()
            for (i in 0 until size) {
                quests[reader.readInteger()] = reader.readByte()
            }
        }
    }

    fun containsFlag(flags: Int, flag: QuestRequirementType): Boolean {
        return flags and flag.value == flag.value
    }
}